js/map.js

```diff
@@ -188,23 +219,135 @@ window.map = (function () {
+  }
 
-    if (mapElement.querySelector('.popup')) {
-      document.querySelector('.popup__close').removeEventListener('click', closeCardHandler);
-      document.removeEventListener('keydown', escPressCloseCardHandler);
-      removeCard();
-    }
+  function filter() {
+    var type = filterTypeElement.value;
+    var price = filterPriceElement.value;
+    var rooms = filterRoomsElement.value;
+    var guests = filterGuestsElement.value;
+    var featuresArray = [];
 
-    Array.prototype.forEach.call(mapPins, function (elem) {
-      elem.parentElement.removeChild(elem);
+    Array.prototype.forEach.call(housingFeaturesCheckboxes, function (elem) {
+      if (elem.checked) {
+        featuresArray.push(elem.value);
+      }
     });
+
+    var filteredPropertyType;
+    var filteredPropertyPrice;
+    var filteredPropertyRooms;
+    var filteredPropertyGuests;
+    var filteredPropertyFeatures;
+
+    if (filteredData.length > 0) {
+      filteredData.length = 0;
+    }
+
+    // Property Filter
+    if (type === 'any') {
+      filteredPropertyType = data.slice();
+    } else {
+      filteredPropertyType = data.filter(function (ad) {
+        return ad.offer.type === type;
+      });
+    }
+
+    // Price Filter
+    if (price === 'any') {
+      filteredPropertyPrice = filteredPropertyType.slice();
+    } else {
+      filteredPropertyPrice = filteredPropertyType.filter(function (ad) {
+        var retProperty = {};
+        switch (price) {
+          case 'middle':
+            retProperty = ad.offer.price >= 10000 && ad.offer.price <= 50000;
+            break;
+          case 'low':
+            retProperty = ad.offer.price < 10000;
+            break;
+          case 'high':
+            retProperty = ad.offer.price > 50000;
+            break;
+        }
+
+        return retProperty;
+      });
+    }
+
+    // Rooms Filter
+    if (rooms === 'any') {
+      filteredPropertyRooms = filteredPropertyPrice.slice();
+    } else {
+      filteredPropertyRooms = filteredPropertyPrice.filter(function (ad) {
+        return ad.offer.rooms === +rooms;
+      });
+    }
+
+    // Guests Filter
+    if (guests === 'any') {
+      filteredPropertyGuests = filteredPropertyRooms.slice();
+    } else {
+      filteredPropertyGuests = filteredPropertyPrice.filter(function (ad) {
+        if (+guests === 0 && ad.offer.guests > 2) {
+          return ad;
+        }
+        return ad.offer.guests === +guests;
+      });
+    }
+
+    // Features Filter
+    if (featuresArray.length <= 0) {
+      filteredPropertyFeatures = filteredPropertyGuests.slice();
+    } else {
+      filteredPropertyFeatures = filteredPropertyGuests.filter(function (ad) {
+        return featuresArray.every(function (elem) {
+          return ad.offer.features.indexOf(elem) >= 0;
+        });
+      });
+    }
+
+    filteredData = filteredPropertyFeatures.slice();
   }
 
+
+  filterTypeElement.addEventListener('change', function () {
+    window.debaunce(applyFilter);
```

Comment:
(Optional) can you create a debounce method to which you can pass a time interval as a second argument and if it's not passed then you use the default?

Reply:
done